docs/howtos/ensembling.rst

```diff
@@ -0,0 +1,297 @@
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Conv(features=32, kernel_size=(3, 3))(x)
+      x = nn.relu(x)
+      x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+      x = nn.Conv(features=64, kernel_size=(3, 3))(x)
+      x = nn.relu(x)
+      x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+      x = x.reshape((x.shape[0], -1))  # flatten
+      x = nn.Dense(features=256)(x)
+      x = nn.relu(x)
+      x = nn.Dense(features=10)(x)
+      x = nn.log_softmax(x)
+      return x
+
+  def get_datasets():
+    """Load MNIST train and test datasets into memory."""
+    ds_builder = tfds.builder('mnist')
+    ds_builder.download_and_prepare()
+    train_ds = tfds.as_numpy(ds_builder.as_dataset(split='train', batch_size=-1))
+    test_ds = tfds.as_numpy(ds_builder.as_dataset(split='test', batch_size=-1))
+    train_ds['image'] = jnp.float32(train_ds['image']) / 255.
+    test_ds['image'] = jnp.float32(test_ds['image']) / 255.
+    return train_ds, test_ds
+
+  def onehot(labels, num_classes=10):
+    x = (labels[..., None] == jnp.arange(num_classes)[None])
+    return x.astype(jnp.float32)
+
+
+  def cross_entropy_loss(logits, labels):
+    return -jnp.mean(jnp.sum(onehot(labels) * logits, axis=-1))
+
+
+  def compute_metrics(logits, labels):
+    loss = cross_entropy_loss(logits, labels)
+    accuracy = jnp.mean(jnp.argmax(logits, -1) == labels)
+    metrics = {
+        'loss': loss,
+        'accuracy': accuracy,
+    }
+    return metrics
+
+Parallel functions
+--------------------------------
+
+We start by creating a parallel version of ``get_initial_params``, which
+retrieves the initial parameters of the models. We do this using `jax.pmap`_.
+The effect of "pmapping" a function is that it will compile the function with
+XLA (similar to `jax.jit`_), but execute it in parallel on XLA devices (e.g., 
+GPUs/TPUs).
+
+.. codediff::
+  :title_left: Single-model
+  :title_right: Ensemble
+
+  @jax.jit #!
+  def get_initial_params(key):
+    init_val = jnp.ones((1, 28, 28, 1), jnp.float32)
+    initial_params = CNN().init(key, init_val)['params']
+    return initial_params
+
+  ---
+  @jax.pmap #!
+  def get_initial_params(key):
+    init_val = jnp.ones((1, 28, 28, 1), jnp.float32)
+    initial_params = CNN().init(key, init_val)['params']
+    return initial_params
+
+Note that for the single-model code above, we use `jax.jit`_ to lazily model 
+(see `Module.init`_'s documentation for more details). For the ensembling
+case, `jax.pmap`_ will map over the first axis of the provided argument ``key``
+by default, so we should make sure that we provide one key for each device when
+we call this function later on.
+
+Next we simply do the same for the functions ``create_optimizer``, 
+``train_step``, and ``eval_step``. We also make a minor change to 
+``eval_model``, which ensures the metrics are used correctly in the parallel
+setting.
+
+.. codediff::
+  :title_left: Single-model
+  :title_right: Ensemble
+
+  # #!
```

Comment:
Is the first `#` necessary? Why not just `#!`?

Reply:
It's a limitation of the current directive, #! will be rewritten to an empty line, and `codeblock` will ignore that. So we need some content on the first line to make sure it aligns nicely with the RHS.

Reply:
Ah I see fine for now. It'll become part of the folklore for HOWTO reviewers ;)